### ISSUE TITLE
Testing Thespian and facing a couple of issues

### ISSUE THREAD

**Issue by samlesu** [godaddy/Thespian#33]
Thanks for releasing this great and promising project. We are still testing and doing some tests with Thespian and we are facing a couple of issues. We would like to know if we are using a good pattern.

## Our current architecture.

We are using multiple Actor Systems, one master and several workers. Each of the actor system is located on different hosts and we are using the _multiprocTCPBase_ base system.

We have two kind of actors:
- a **TaskManager** located on the **master**. This actor is responsible of maintaining a pool of connected **TaskExecutor**s and scheduling some tasks. 
- a **TaskExecutor** located on the **worker**. It is responsible of executing tasks.

When starting the **master** actor system, a **TaskManager** actor is created. The **TaskManager** is then listening at convention updates. On the **worker** side, the actor system is connecting to the master address and no actor is created.
When the **TaskManager** is notified that a new actor system is joining the convention, it is starting a new **TaskExecutor** on the remote actor system (this is done by using unique capabilities).
To unregister a **TaskExecutor**, we are using the _ChildActorExited_ message. It is working as expected when a remote actor system is performing a graceful shutdown.

## Maintaining a list of active workers.

We are facing some issues to maintain a list of active **TaskExecutor**s. When a **worker** is killed, no convention update is received by the **master**, and no _ChildActorExited_ message is received. 
How could we detect that a remote actor system has been killed and that a child actor no longer exist? Could we list all the child actors on a remote actor system?

## False convention updates.

From time to time, we received some false convention updates. From time to time, the master receives convention update regarding a worker leaving and joining straight after the convention. Is it an expected behaviour?

## Creating other processes.

As said before, the role of the **TaskExecutor** is to execute tasks. Tasks are executed asynchronously by creating new processes (to avoid blocking the current actor system). According to the documentation it could result in unexpected behaviour. We noticed that while running a background task in a separate process most of the messages are delayed. 
How could we deal with it?

Thanks for reading me and thanks again for your time.

**Comment by kwquick** [comment]
I'm glad you are finding Thespian useful, and thank's for the kudos!

The architectural pattern you have chosen is one that I would recommend: we also use this pattern and we find that it works quite well overall (our environment is 9000+ remote systems, with typically 3-8 actors on each remote).

## Worker Liveness
If I understand correctly, your first issue is that if a **worker** system is shutdown abruptly and cannot perform the graceful shutdown that the master system has no indication of this.  The convention membership does employ period keep-alive messages to detect when a remote is no longer responding, but this is set to missing 3 checkins that currently occur at about a 7m22s period, so you should receive the indications you are expecting, but only after about 22-23 minutes (the idea being that this is an unusual scenario so the keep-alive should not be unduly burdensome to the system).  Please let me know if this is not occurring, as this should be fixed.  I can see a couple of options for getting a faster indication:
   

1. Implement your own keep-alive (between **TaskManager** and **TaskExecutor**) with a shorter period and use that in addition to the _ChildActorExited_ messages,
2. Update Thespian to make the period and retry count a configurable option.  This would involve modifications to the `thespian/system/admin/convention.py` module to override the `CONVENTION_REREGISTRATION_PERIOD` and `CONVENTION_REGISTRATION_MISS_MAX` values with optional capabilities in the `ConventionMemberData.__init__()`.

I don't think option #2 is unreasonable, but I would like to confirm that this would resolve your issues first and I would like to think about the safeties involved in ensuring those don't become too intrusive,  You may want to manually adjust those constants to see if that achieves the results you would like to see and let me know what your results are.

## False Convention Updates

The remote system and local system use the keep-alive messages to try to determine continuity of connection.  If the remote system is restarted, then when it first checks back in to the convention, there is functionality to determine that this is a _new_ instance of the instead of just another checkin from a still-running instance.  In the case that the local system thinks that a new instance of the remote system is checking in, it will generate back-to-back convention updates to notification subscribers indicating that the remote has left and then rejoined; this is also an indication to the subscriber that any remote state must be re-established.

In addition, if the remote system's capabilities change (via an `updateSystemCapabilities()` call) then there will be a new registration and notification so that the new capabilities are observed.

If you think you are getting spurious notifications outside of the above scenarios I would like to work with you to understand the scenario better to find out why these are happening.  Also note that there is a bugfix release I expect to have out today or tomorrow and while none of the fixes are directly related to this behavior, you may see some improvement through side effects depending on your configuration.

## Creating Other Processes

In general, you can think of the `Actor` base class as a wrapper around your `receiveMessage()` method, where the wrapper for the multiprocTCPBase does roughly the following pseudo-code:

    while True:
        r = select.select()
        if r is sendable:
            socket.send(...)
        if r is receivable:
            message = socket.read()
            self.receiveMessage(message, ...)
            if isinstance(message, ActorExitRequest):
                sys.exit()

That's hugely oversimplified, but you can see that if the `receiveMessage()` method blocks, then this core loop does not receive additional messages or send portions of outbound messages (sends are asynchronous).  This helps to explain the message delays you have observed.

There are two approaches I could suggest for your scenario.

1. If the task to be executed is being done within Python and has a socket (or linux file descriptor) that can be selected on so that it could be run in a non-blocking manner, you could use the Thespian Watch extension: http://thespianpy.com/using.html#sec-4-8

2. If the task is invoking an externally blocking operation then I would recommend using a **taskrunner** actor specifically to perform that operation.  The **TaskExecutor** can create a **taskrunner** for each task execution (and the **taskrunner** can exit on completion); the **TaskExecutor** can be written to be aware that **taskrunner** actor instances may be unresponsive, but the **TaskExecutor** still remains responsive and able to communicate with the **TaskManager** without blocking.

Hopefully the above helps, and again, thanks for your interest in Thespian and your assistance and questions in helping to improve it.

**Comment by samlesu** [comment]
Thank you for your quick and detailed answer. Things are much more clearer now. We will give you our feedback after doing some changes.

**Comment by kwquick** [comment]
Closing this issue because of lack of updates.  Please feel free to re-open or create another issue if you still have questions or problems.
